summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala

```diff
@@ -1,102 +1,238 @@
+    node: StormNode,
+    provider: ComponentProvider
+  ): Topology = {
+    val component = provider.createSingle(identity[Item[T]])
+    val (componentId, topologyWithComponent) = topology.withComponent(getNodeName(node), component)
+    registerItemEdges[T](topologyWithComponent, node, componentId)
   }
 
-  private def registerIncomingEdges(topology: Topology, node: StormNode, edgeType: EdgeType[_]): Topology = {
-    val nodeId: ReceivingId[Any] = getId(node).asInstanceOf[Topology.ReceivingId[Any]]
-    stormDag.dependenciesOf(node).foldLeft(topology) { (current, upstreamNode) =>
-      current.withEdge(Topology.Edge[Any](
-        getId(upstreamNode),
-        edgeType.asInstanceOf[EdgeType[Any]],
-        nodeId))
-    }
+  private def registerKeyValueComponent[K, V](
+    topology: Topology,
+    node: StormNode,
+    provider: ComponentProvider
+  ): Topology = outgoingSummersProps(node) match {
+    case Some(props) if props.allSummers =>
+      // Use Aggregated if possible.
+      provider.createAggregated[K, V](props.batcher, props.shards, props.semigroup
+        .asInstanceOf[Semigroup[V]]) match {
+        case Some(component) =>
+          val (componentId, topologyWithComponent) = topology.withComponent(getNodeName(node), component)
+          registerAggregatedEdges[K, V](topologyWithComponent, node, componentId)
+        case None =>
+          // Fallback to Sharded.
+          registerShardedKeyValue[K, V](topology, node, provider, props.batcher, props.shards)
+      }
+    case Some(props) =>
+      registerShardedKeyValue[K, V](topology, node, provider, props.batcher, props.shards)
+    case None =>
+      throw new IllegalStateException("Shouldn't happen, trying to register component which emits " +
+        "keyed values and don't have downstream summers.")
   }
 
-  private def getId(node: StormNode): Topology.EmittingId[_] = node match {
-    case sourceNode: SourceNode[Storm] => getSourceId(sourceNode)
-    case flatMapNode: FlatMapNode[Storm] => getFlatMapId(flatMapNode)
-    case summerNode: SummerNode[Storm] => getSummerId(summerNode)
+  private def registerShardedKeyValue[K, V](
+    topology: Topology,
+    node: StormNode,
+    provider: ComponentProvider,
+    batcher: Batcher,
+    shards: KeyValueShards
+  ): Topology = {
+    val component = provider.createSingle[(K, V), Sharded[K, V]](wrapSharded(batcher, shards))
+    val (componentId, topologyWithComponent) = topology.withComponent(getNodeName(node), component)
+    registerShardedEdges[K, V](topologyWithComponent, node, componentId)
   }
 
-  private def getSummerId(node: SummerNode[Storm]): Topology.BoltId[_, _] =
+  private def registerItemEdges[T](
+    topology: Topology,
+    source: StormNode,
+    sourceId: Topology.EmittingId[Item[T]]
+  ): Topology = stormDag.dependantsOf(source).foldLeft(topology) {
+    case (currentTopology, downstreamNode: FlatMapNode[Storm]) =>
+      currentTopology.withEdge(Edges.shuffleItemToItem[T](
```

Comment:
I'm not sure I understand about which expression exactly you are saying. IMO `edge` variable extraction doesn't improve readability and add a little bit of noise here. If separate variable here makes understanding of code here much easier - I'll change it.